kubernetes/controller/internal/configuration/config.go

```diff
@@ -0,0 +1,123 @@
+// Package configuration provides functionality to load and manage OCM configurations
+// from Kubernetes resources (Secrets and ConfigMaps).
+package configuration
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	genericv1 "ocm.software/open-component-model/bindings/go/configuration/generic/v1/spec"
+	"ocm.software/open-component-model/bindings/go/runtime"
+	"ocm.software/open-component-model/kubernetes/controller/api/v1alpha1"
+)
+
+// GetConfigFromSecret extracts and decodes OCM configuration from a Kubernetes Secret.
+// It looks for configuration data under the OCMConfigKey.
+func GetConfigFromSecret(secret *corev1.Secret) (*genericv1.Config, error) {
+	data, ok := secret.Data[v1alpha1.OCMConfigKey]
+	if !ok || len(data) == 0 {
+		return nil, nil
+	}
+
+	var cfg genericv1.Config
+	if err := genericv1.Scheme.Decode(bytes.NewReader(data), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to decode ocm config from secret %s/%s: %w",
+			secret.Namespace, secret.Name, err)
+	}
+
+	return &cfg, nil
+}
+
+// GetConfigFromConfigMap extracts and decodes OCM configuration from a Kubernetes ConfigMap.
+// It looks for configuration data under the OCMConfigKey.
+func GetConfigFromConfigMap(configMap *corev1.ConfigMap) (*genericv1.Config, error) {
+	data, ok := configMap.Data[v1alpha1.OCMConfigKey]
+	if !ok || len(data) == 0 {
+		return nil, nil
+	}
+
+	var cfg genericv1.Config
+	if err := genericv1.Scheme.Decode(strings.NewReader(data), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to decode ocm config from configmap %s/%s: %w",
+			configMap.Namespace, configMap.Name, err)
+	}
+
+	return &cfg, nil
+}
+
+// GetConfigFromObject extracts configuration from either a Secret or ConfigMap.
+func GetConfigFromObject(obj client.Object) (*genericv1.Config, error) {
+	switch o := obj.(type) {
+	case *corev1.Secret:
+		return GetConfigFromSecret(o)
+	case *corev1.ConfigMap:
+		return GetConfigFromConfigMap(o)
+	default:
+		return nil, fmt.Errorf("unsupported configuration object type: %T", obj)
+	}
+}
+
+// LoadConfigurations loads OCM configurations from a list of OCMConfiguration references.
+// It fetches the referenced Secrets/ConfigMaps from the cluster and extracts their configuration.
+// All configurations are merged using FlatMap into a single Config.
+func LoadConfigurations(ctx context.Context, k8sClient client.Reader, namespace string, ocmConfigs []v1alpha1.OCMConfiguration) (*genericv1.Config, error) {
+	if len(ocmConfigs) == 0 {
+		return &genericv1.Config{}, nil
+	}
+
+	// Best effort sort of the configs to ensure they are always processed in the same order.
+	sort.Slice(ocmConfigs, func(i, j int) bool {
+		key1 := ocmConfigs[i].Name + "." + ocmConfigs[i].Namespace
```

Comment:
I think you could sort by UID no?

Reply:
it doesn't have one. 🤔

Reply:
huh?! every object in k8s has a uid no?

Reply:
Okay, need to figure out how to deal with this effectively. This is important and critical code section called many times. Hashing and sorting all the time every config and individually getting all the items might be a bit too slow.

Reply:
Okay, I _think_ I parallelized most of the critical paths here. It can be improved of course ( constant for a semaphore right now ), but I want your first eyes on it.